opentelemetry-sdk/src/metrics/internal/histogram.rs

```diff
@@ -1,14 +1,49 @@
-use std::{collections::HashMap, sync::Mutex, time::SystemTime};
+use std::collections::HashSet;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::{sync::Mutex, time::SystemTime};
 
+use crate::metrics::data::HistogramDataPoint;
 use crate::metrics::data::{self, Aggregation, Temporality};
-use crate::{metrics::data::HistogramDataPoint, metrics::AttributeSet};
 use opentelemetry::KeyValue;
-use opentelemetry::{global, metrics::MetricsError};
 
-use super::{
-    aggregate::{is_under_cardinality_limit, STREAM_OVERFLOW_ATTRIBUTE_SET},
-    Number,
-};
+use super::Number;
+use super::{AtomicTracker, AtomicallyUpdate, Operation, ValueMap};
+
+struct HistogramUpdate;
+
+impl Operation for HistogramUpdate {
+    fn update_tracker<T: Default, AT: AtomicTracker<T>>(tracker: &AT, value: T, index: usize) {
+        tracker.update_histogram(index, value);
+    }
+}
+
+struct HistogramTracker<T> {
+    buckets: Mutex<Buckets<T>>,
+}
+
+impl<T: Number<T>> AtomicTracker<T> for HistogramTracker<T> {
+    fn update_histogram(&self, index: usize, value: T) {
+        let mut buckets = match self.buckets.lock() {
+            Ok(guard) => guard,
+            Err(_) => return,
+        };
+
+        buckets.bin(index, value);
+        buckets.sum(value);
```

Comment:
One difference now is that we would always update the sum. Previously, this was happening only when `self.record_sum` is `true`. I think we are not required to provide `record_sum` as a configuration option in the first place.

Reply:
spec:

> 
Arithmetic sum of Measurement values in population. This SHOULD NOT be collected when used with instruments that record negative measurements (e.g. UpDownCounter or ObservableGauge).

This scenario occurs only if a View is used to aggregate UpDownCounter/Gauge as Histogram. Lets address this (if we chose to), in a future PR.

Reply:
Additionally, I am unsure if SUM should be not reported, when -ve measurements are provided, which are currently not blockeable at API level for f64...

Proto wording is not super clear to me...Lets follow up later.
https://github.com/open-telemetry/opentelemetry-proto/blob/v0.9.0/opentelemetry/proto/metrics/v1/metrics.proto#L538-L543